torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -870,6 +872,123 @@ def _check_supported_callable(fn_var):
         )
 
 
+class AssociativeScanHigherOrderVariable(TorchHigherOrderOperatorVariable):
+    @raise_hard_error_if_graph_break(
+        reason="associative_scan must be captured completely with torch.compile."
+    )
+    def call_function(
+        self, tx, args: List[VariableTracker], kwargs: Dict[str, VariableTracker]
+    ) -> VariableTracker:
+        from . import NestedUserFunctionVariable, UserFunctionVariable
+        from .builder import wrap_fx_proxy
+
+        args, kwargs = LazyVariableTracker.realize_all((args, kwargs))
+
+        def arg_extractor(input, dim, combine_fn):
+            return input, dim, combine_fn
+
+        input, dim, combine_fn = arg_extractor(*args, **kwargs)
+
+        assert isinstance(
+            combine_fn,
+            (
+                UserFunctionVariable,
+                NestedUserFunctionVariable,
+            ),
+        )
+
+        # operands
+        if input.python_type() != torch.Tensor:
+            unimplemented(
+                f"Expected input to be a tensor but got {input.python_type()}",
+            )
+
+        input_meta = input.as_proxy().node.meta["example_value"]
+        with tx.fake_mode:
+            example_vals = [
+                torch.full(
+                    (), float("nan"), device=input_meta.device, dtype=input_meta.dtype
+                )
+                for _ in range(2)
+            ]
+
+        # Trace the subgraph
+        # NOTE: We don't use speculate_subgraph as it expects subgraph arguments to be
+        # variables in the current graph.
+        with tx.output.subtracer(self.value, prior_tracer=None) as subtracer:
+            sub_args = [
+                wrap_fx_proxy(
+                    tx=tx,
+                    proxy=subtracer.create_graph_input(name),
+                    example_value=example,
+                )
+                for name, example in zip("ab", example_vals)
+            ]
+
+            combine_result = combine_fn.call_function(tx, sub_args, {})
+            combine_result_proxies = combine_result.as_proxy()
+            combine_result_proxies = pytree.tree_map(
+                subtracer.maybe_lift_tracked_freevar_to_input, combine_result_proxies
+            )
+
+            tx.output.create_node(
+                "output",
+                "output",
+                (subtracer.create_arg((combine_result_proxies,))),
+                {},
+            )
+
+        combine_graph = subtracer.graph
+        combine_graph.lint()
+
+        if subtracer.lifted_freevars:
+            unimplemented(
+                f"Combine fn had unexpected freevars: {subtracer.lifted_freevars}"
+            )
```

Comment:
Can we graphbreak at all here? Probably better throw a hard error. Same below.

Reply:
This function is decorated with `@raise_error_if_graph_break` so this does result in a hard error.

Reply:
Why not simply raise an error directly?

Reply:
This is consistent with how it's written in cond and while_loop. I suppose one reason to do it this way is all errors will give a similar error message no matter where in the graph capture process it errors.